docs/markdown/Rewriter.md

```diff
@@ -0,0 +1,220 @@
+---
+short-description: Automatic modification of the build system files
+...
+
+# Meson file rewriter
+
+Since version 0.50.0, meson has the functionality to perform some basic
+modification on the `meson.build` files from the command line. The currently
+supported operations are:
+
+- For build targets:
+  - Add/Remove source files
+  - Add/Remove targets
+  - Modify a select set of kwargs
+  - Print some JSON information
+- For dependencies:
+  - Modify a select set of kwargs
+- For the project function:
+  - Modify a select set of kwargs
+  - Modify the default options list
+
+The rewriter has both, a normal command line interface and a "script mode". The
+normal CLI is mostly designed for everyday use. The "script mode", on the
+other hand, is meant to be used by external programs (IDEs, graphical
+frontends, etc.)
+
+## Using the rewriter
+
+All rewriter functions are accessed via `meson rewrite`. The meson rewriter
+assumes that it is run inside the project root directory. If this isn't the
+case, use `--sourcedir` to specify the actual project source directory.
+
+### Adding and removing sources
+
+The most common operations will probably be the adding and removing of source
+files to a build target. This can be easily done with:
+
+```bash
+meson rewrite target <target name/id> {add/rm} [list of sources]
+```
+
+For instance, given the following example
+
+```meson
+src = ['main.cpp', 'fileA.cpp']
+
+exe1 = executable('testExe', src)
+```
+
+the source `fileB.cpp` can be added with:
+
+```bash
+meson rewrite target testExe add fileB.cpp
+```
+
+After executing this command, the new `meson.build` will look like this:
+
+```meson
+src = ['main.cpp', 'fileA.cpp', 'fileB.cpp']
+
+exe1 = executable('testExe', src)
+```
+
+In this case, `exe1` could also have been used for the target name, since the
+rewriter also takes assignments and internal meson IDs into consideration when
+searching for the target in the `meson.build` files.
+
+For more information see the help output of the rewriter target command.
+
+### Setting the project version
+
+It is also possible to set kwargs of specific functions with the rewriter. The
+general command for setting or removing kwargs is:
+
+```bash
+meson rewriter kwargs {set/delete} <function type> <function ID> <key1> <value1> <key2> <value2> ...
+```
+
+For instance, setting the project version can be achieved with this command:
+
+```bash
+meson rewriter kwargs set project '' version 1.0.0
```

Comment:
This is not a particularly good UX. Magic empty strings are not particularly nice. Could this be changed so that these sorts of strings are not necessary?

Reply:
I would love to do that, but the problem is that I have implemented this with positional arguments. I have already tried making the ID optional but I never managed to get it to work.

I could make a separate shortcut subcommand only for project, but this seems like a hack. The only other option would be to rely on `--` flags and I _really_ wanted to avoid this.

Reply:
This seems mostly in the realm of scripts and IDEs. I don't expect our users to change the version number often via this command. So i'm not sure how bad this is.

Of course we could also define `/` to mean the top level project. I don't think that has a chance of colliding with any likely ways to specify subprojects (that seems the likely use for this parameter in the future)

Reply:
@jpakkane Would you be ok with a `/` or `root` or something similar?

Reply:
Let's use `/`.

Reply:
Done